libraries/eosiolib/varint.hpp

```diff
@@ -427,14 +426,12 @@ struct signed_int {
     template<typename DataStream>
     friend DataStream& operator >> ( DataStream& ds, signed_int& vi ){
       uint32_t v = 0; char b = 0; int by = 0;
-      do {
+      do {                                                      //read 7 bit chunks
          ds.get(b);
          v |= uint32_t(uint8_t(b) & 0x7f) << by;
          by += 7;
       } while( uint8_t(b) & 0x80 );
-      vi.value = ((v>>1) ^ (v>>31)) + (v&0x01);
-      vi.value = v&0x01 ? vi.value : -vi.value;
-      vi.value = -vi.value;
+      vi.value= (v>>1) ^ -(v&1);                                //reverse zigzag encoding
```

Comment:
This seems a little obtuse to me.  The expression -(v&1) is not immediately apparent that it is not undefined behavior and takes some staring at to see that it is fine and what it is doing.  I think it might be better to use 1ull + ~(v&1), both should compile to the same instructions (so no loss of performance, but a little more obvious in my opinion).